### PR TITLE
feat(meta): improve param updating process when the cluster version upgrades

### PR DESCRIPTION
**Pull request by Gun9niR**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

When the RW cluster upgrades to a new version, new parameters may be introduced. They may either be:

- Entirely new params that need to be initialized
- Or newer versions of existing parameters that need to retain the same effect on the cluster.

For the first case, `SystemParamManager` will detect missing params and initialize them with CLI values or default values.

For the second case, a framework is added to conveniently derive the value of new params with existing params. The developer simply needs to override the trait `OverrideFromParams` to define custom rules.

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [x] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: src/common/src/system_param/mod.rs [hunk 1]
```diff
@@ -95,45 +116,50 @@ macro_rules! impl_system_params_to_kv {
...
 macro_rules! impl_system_params_from_kv {
     ($({ $field:ident, $type:ty, $default:expr },)*) => {
-        /// For each field in `SystemParams`, one of these rules apply:
-        /// - Up-to-date: Guaranteed to be `Some`. If it is not present, may try to derive it from previous
-        ///   versions of this field.
-        /// - Deprecated: Guaranteed to be `None`.
-        /// - Unrecognized: Not allowed.
-        pub fn system_params_from_kv(kvs: Vec<(impl AsRef<[u8]>, impl AsRef<[u8]>)>) -> Result<SystemParams> {
+        /// Try to deserialize deprecated fields as well.
+        /// Return error if there are unrecognized fields
```
Comment (BugenZhao) [inline src/common/src/system_param/mod.rs:135]:
```suggestion
        /// Return error if there are unrecognized fields.
```

File: src/common/src/system_param/mod.rs [hunk 2]
```diff
@@ -95,45 +116,50 @@ macro_rules! impl_system_params_to_kv {
...
-            }
-            if !expected_keys.is_empty() {
-                return Err(format!(
-                    "missing system param {:?}",
-                    expected_keys
-                ));
+            });
+            derive_missing_fields(&mut ret);
+            if !kvs.is_empty() {
+                 Err(format!("unrecognized system params {:?}", kvs))
```
Comment (BugenZhao) [inline src/common/src/system_param/mod.rs:159]:
```suggestion
                Err(format!("unrecognized system params {:?}", kvs))
```

File: src/common/src/system_param/mod.rs [hunk 3]
```diff
@@ -95,45 +116,50 @@ macro_rules! impl_system_params_to_kv {
...
-                $(key_of!($field),)*
-            ]
-            .iter()
-            .cloned()
-            .collect();
-            for (k, v) in kvs {
+            kvs.retain(|(k,v)| {
                 let k = std::str::from_utf8(k.as_ref()).unwrap();
                 let v = std::str::from_utf8(v.as_ref()).unwrap();
                 match k {
```
Comment (BugenZhao) [inline src/common/src/system_param/mod.rs:145]: We may only let the macro generate this `match` statement and define the function normally, so it'll be more friendly to editing.

File: src/common/src/system_param/mod.rs [hunk 4]
```diff
@@ -170,6 +196,39 @@ macro_rules! impl_default_validation_on_set {
...
+// semantic change, where a new parameter has to be introduced. When the cluster upgrades to a newer
+// version, we need to ensure the effect of the new parameter is equal to its older versions.
+//
+// For example, if you had `interval_sec` and now you want finer granularity, you can introduce a
+// new param `interval_ms` and try to derive it from `interval_sec` by overriding `FromParams` trait
+// in `OverrideFromParams`:
+//
+// ```
+// impl FromParams for OverrideFromParams {
+//     fn interval_ms(params: &mut SystemParams) -> Option<u64> {
```
Comment (BugenZhao) [inline src/common/src/system_param/mod.rs:209]: Maybe it's better to take an immutable ref here?

File: src/common/src/system_param/mod.rs [hunk 5]
```diff
@@ -170,6 +196,39 @@ macro_rules! impl_default_validation_on_set {
     }
 }
 
+// Define rules to derive a parameter from others. This is useful for parameter type change or
```
Comment (BugenZhao) [inline src/common/src/system_param/mod.rs:199]:
```suggestion
/// Define rules to derive a parameter from others. This is useful for parameter type change or
```

File: src/common/src/system_param/mod.rs [hunk 6]
```diff
@@ -40,19 +42,24 @@ macro_rules! for_all_undeprecated_params {
...
-// Only includes deprecated params. Used to define key constants.
+// Includes all params.
 // Macro input is { field identifier, type, default value }
-macro_rules! for_all_deprecated_params {
+macro_rules! for_all_params {
     ($macro:ident) => {
-        $macro! {}
+        for_all_undeprecated_params!(
+            $macro /* Define future deprecated params here, such as
+                    * ,{ backup_storage_directory, String, "backup".to_string() } */
```
Comment (BugenZhao) [inline src/common/src/system_param/mod.rs:56]: This really works but seems a little bit strange to pass deprecated params to the macro of `undeprecated_params`. 🤣

**Review by BugenZhao — COMMENTED** [review]
Rest LGTM.

File: src/common/src/system_param/mod.rs (hunk 6 above)
Comment (Gun9niR) [reply src/common/src/system_param/mod.rs:56]: I really can't think of another way to put two `for_all_xxx` mcros together 😢

File: src/common/src/system_param/mod.rs (hunk 3 above)
Comment (Gun9niR) [reply src/common/src/system_param/mod.rs]: The difficulty is that the macro to generate `match` needs to capture `k, v, ret`. In [this commit](https://github.com/risingwavelabs/risingwave/pull/8260/commits/f81000a01206110b94c17047562a3984c01e65ee), I modified `for_all_params` and `for_all_undeprecated_params` to accept custom arguments, but it makes the macro definition and usage more complex.

Comment (BugenZhao) [reply src/common/src/system_param/mod.rs]: You might be right. 🥵 Let's prefer simplicity.

**Pull request merged**